File: its-my-training-day/leetcode-top-75/questions/dynamic-programming/dp-2d/longest-common-subsequence/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def longest_common_subsequence(text1: str, text2: str) -> int:
    n = len(text1)
    m = len(text2)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    i = 1
    while i <= n:
        j = 1
        while j <= m:
            if text1[i - 1] == text2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
            j += 1
        i += 1
    return dp[n][m]

File: its-my-training-day/leetcode-top-75/questions/dynamic-programming/dp-2d/longest-common-subsequence/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const longestCommonSubsequence = (text1: string, text2: string): number => {
  const n = text1.length;
  const m = text2.length;
  const dp = new Array(n + 1);
  for (let i = 0; i <= n; i += 1) {
    dp[i] = new Array(m + 1).fill(0);
  }
  for (let i = 1; i <= n; i += 1) {
    for (let j = 1; j <= m; j += 1) {
      if (text1[i - 1] === text2[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
      else dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
    }
  }
  return dp[n][m];
};
